### PR TITLE
Implement strict integer operations that panic on overflow

### PR DESCRIPTION
**Pull request by rmehri01**
This PR implements the first part of the ACP for adding panic on overflow style arithmetic operations (https://github.com/rust-lang/libs-team/issues/270), mentioned in #116064.

It adds the following operations on both signed and unsigned integers:

- `strict_add`
- `strict_sub`
- `strict_mul`
- `strict_div`
- `strict_div_euclid`
- `strict_rem`
- `strict_rem_euclid`
- `strict_neg`
- `strict_shl`
- `strict_shr`
- `strict_pow`

Additionally, signed integers have:

- `strict_add_unsigned`
- `strict_sub_unsigned`
- `strict_abs`

And unsigned integers have:

- `strict_add_signed`

The `div` and `rem` operations are the same as normal division and remainder but are added for completeness similar to the corresponding `wrapping_*` operations.

I'm not sure if I missed any operations, I basically found them from the `wrapping_*` and `checked_*` operations on both integer types.

### CODE REVIEW

**Comment by asquared31415** [comment]
All of the exposed methods should probably be `#[track_caller]` to improve the panic locations.

**Comment by eduardosm** [comment]
cc @RalfJung (I didn't know if you were aware of this PR)

File: library/core/src/num/int_macros.rs [hunk 1]
```diff
@@ -454,6 +454,39 @@ macro_rules! int_impl {
...
+        /// ```
+        /// #![feature(strict_overflow_ops)]
+        #[doc = concat!("assert_eq!((", stringify!($SelfT), "::MAX - 2).strict_add(1), ", stringify!($SelfT), "::MAX - 1);")]
+        /// ```
+        ///
+        /// ```should_panic
+        /// #![feature(strict_overflow_ops)]
+        #[doc = concat!("let _ = (", stringify!($SelfT), "::MAX - 2).strict_add(3);")]
+        /// ```
+        #[unstable(feature = "strict_overflow_ops", issue = "116064")]
```
Comment (RalfJung) [inline library/core/src/num/int_macros.rs:479]: It's probably better to create a new tracking issue for this, with the tracking issue template.

Comment (rmehri01) [reply library/core/src/num/int_macros.rs:479]: Good point, I created #118260.

File: library/core/src/num/int_macros.rs [hunk 2]
```diff
@@ -454,6 +454,39 @@ macro_rules! int_impl {
...
         }
 
+        /// Strict integer addition. Computes `self + rhs`, panicking
+        /// if overflow occurred.
+        ///
+        /// # Panics
+        ///
+        /// ## Overflow behavior
+        ///
+        /// This function will always panic on overflow, regardless of if overflow checks are enabled.
```
Comment (m-ou-se) [inline library/core/src/num/int_macros.rs:464]:
```suggestion
        /// This function will always panic on overflow, regardless of whether overflow checks are enabled.
```

File: library/core/src/num/overflow_panic.rs
[the remark this answers is not in the document]
Comment (RalfJung) [reply library/core/src/num/overflow_panic.rs:1]: I think `description` is used in a bunch of places where we really need a string literal, e.g. for MIR dumping.

But if there's some way to reduce the redundancy here, I'm all for it. This duplication of the panic message in 3 or 4 places has bothered me for a while, I just never found a great way to fix it.

**Comment by RalfJung** [comment]
@m-ou-se just a reminder that there's a PR waiting for review here :)

**Pull request merged**